File: OpenVibbleApp/BridgeAppModel.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation
import UIKit
import Combine
import CoreBluetooth
import BridgeRuntime
import NUSPeripheral
import BuddyProtocol
import BuddyStorage
import BuddyStats
import BuddyPersona

@MainActor
final class BridgeAppModel: ObservableObject {
    @Published private(set) var snapshot: BridgeSnapshot = .empty
    @Published private(set) var prompt: PromptRequest?
    @Published private(set) var transfer: TransferProgress = .idle
    @Published private(set) var connectionState: NUSConnectionState = .stopped
    @Published private(set) var bluetoothStateNote: String = "蓝牙状态未知"
    @Published private(set) var bluetoothPowerState: CBManagerState = .unknown
    @Published private(set) var advertisingNote: String = "未广播"
    @Published private(set) var activeDisplayName: String = "Claude"
    @Published private(set) var diagnosticLogs: [String] = []
    @Published private(set) var recentEvents: [String] = []
    /// Client-side accumulation of heartbeat `entries`. The bridge replaces
    /// its list every heartbeat (see REFERENCE.md — "capped to a few"), so
    /// `snapshot.entries` by itself never grows past ~3 items. We merge each
    /// heartbeat into a longer scrollback here, newest first, deduping
    /// against a sliding window so a line that stays in the bridge's top-N
    /// across several heartbeats doesn't get re-prepended.
    @Published private(set) var parsedEntries: [String] = []
    private static let parsedEntriesMax = 200
    private static let parsedEntriesDedupWindow = 16
    @Published private(set) var lastInstalledCharacter: String?
    @Published private(set) var recentLevelUp: Bool = false
    @Published private(set) var lastQuickApprovalAt: Date?
    /// Bumped each time a heartbeat with `completed: true` arrives. Drives a
    /// 3-second celebrate animation via `PersonaController` — matches the h5
    /// demo's `triggerOneShot("celebrate", 3000)` and the firmware's
    /// `P_CELEBRATE` state derivation.
    @Published private(set) var lastCompletedAt: Date?
    /// True once the user has responded to the current prompt but the desktop
    /// has not yet cleared it from the next heartbeat. Used by the prompt
    /// panel to swap the "A approve / B deny" hint for a "SENT" state so
    /// users see immediate feedback instead of the waited-timer ticking on.
    @Published private(set) var responseSent: Bool = false
    /// Mirrors `BuddyPeripheralService.authorizationState` so views can react
    /// to the user granting/denying the BLE permission.
    @Published private(set) var bluetoothAuthorization: CBManagerAuthorization = CBPeripheralManager.authorization

    private let quickApprovalThreshold: TimeInterval = 5
    private let liveActivityManager = BuddyLiveActivityManager()

    let statsStore: PersonaStatsStore

    private let runtime = BridgeRuntime()
    private let peripheral = BuddyPeripheralService()
    private var cancellables: Set<AnyCancellable> = []
    private var started = false
    private var lastPromptId: String?
    private var lastPromptAt: Date?
    private var statusSampleTimer: Timer?

    var charactersRootURL: URL { runtime.charactersRootURL }

    /// Per-project grouping of the rolling heartbeat entries, used by the
    /// INFO > CLAUDE page. Recomputed on every read — SwiftUI diffs the
    /// resulting list, and the input (`parsedEntries` + `prompt`) already
    /// triggers view invalidation via `@Published`.
    var projects: [ProjectSummary] {
        ProjectSummaryBuilder.build(entries: parsedEntries, hasPrompt: prompt != nil)
    }

    init(statsStore: PersonaStatsStore = PersonaStatsStore()) {
        self.statsStore = statsStore
        runtime.onCharacterInstalled = { [weak self] name in
            Task { @MainActor [weak self] in
                self?.lastInstalledCharacter = name
                self?.recordEvent("系统 已安装宠物：\(name)")
            }
        }

        runtime.onSpeciesChanged = { [weak self] idx in
            Task { @MainActor [weak self] in
                if idx == PersonaSpeciesCatalog.gifSentinel {
                    self?.recordEvent("系统 切换宠物 → GIF")
                } else if let name = PersonaSpeciesCatalog.name(at: idx) {
                    self?.recordEvent("系统 切换宠物 → \(name) (idx=\(idx))")
                }
            }
        }

        runtime.onTaskCompleted = { [weak self] in
            Task { @MainActor [weak self] in
                self?.lastCompletedAt = Date()
            }
        }

        enableBatteryMonitoring()

        peripheral.$connectionState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.connectionState = state
                self?.pushLiveActivity()
            }
            .store(in: &cancellables)

        // React to the user toggling the Live Activity preference in Settings:
        // flipping it off mid-session should tear down any running activity
        // immediately instead of waiting for the next heartbeat.
        NotificationCenter.default.publisher(for: UserDefaults.didChangeNotification)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in
                self?.pushLiveActivity()
            }
            .store(in: &cancellables)

        peripheral.$bluetoothStateNote
            .receive(on: DispatchQueue.main)
            .sink { [weak self] note in
                self?.bluetoothStateNote = note
            }
            .store(in: &cancellables)

        peripheral.$bluetoothPowerState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.bluetoothPowerState = state
            }
            .store(in: &cancellables)

        peripheral.$advertisingNote
            .receive(on: DispatchQueue.main)
            .sink { [weak self] note in
                self?.advertisingNote = note
            }
            .store(in: &cancellables)

        peripheral.$diagnostics
            .receive(on: DispatchQueue.main)
            .sink { [weak self] logs in
                self?.diagnosticLogs = logs
            }
            .store(in: &cancellables)

        peripheral.$authorizationState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] auth in
                self?.bluetoothAuthorization = auth
            }
            .store(in: &cancellables)

        peripheral.onLineReceived = { [weak self] line in
            guard let self else { return }
            self.recordEvent("接收  \(line)")
            let outbound = self.runtime.ingestLine(line)
            self.refreshFromRuntime()
            for response in outbound {
                self.recordEvent("发送  \(response.trimmingCharacters(in: .whitespacesAndNewlines))")
                _ = self.peripheral.sendLine(response)
            }
        }

        subscribeLiveActivityDecisions()
    }

    /// Listens for Approve/Deny decisions posted by the Live Activity AppIntent
    /// (running in the widget process). We register a Darwin notification so we
    /// hear it regardless of app foreground state, then drain the shared
    /// `UserDefaults` record on the main actor.
    private func subscribeLiveActivityDecisions() {
        let rawName = LiveActivitySharedStore.decisionChangedDarwinName
        let observer = Unmanaged.passUnretained(self).toOpaque()
        CFNotificationCenterAddObserver(
            CFNotificationCenterGetDarwinNotifyCenter(),
            observer,
            { _, _, _, _, _ in
                Task { @MainActor in
                    NotificationCenter.default.post(
                        name: .liveActivityDecisionReceived,
                        object: nil
                    )
                }
            },
            rawName as CFString,
            nil,
            .deliverImmediately
        )

        NotificationCenter.default.publisher(for: .liveActivityDecisionReceived)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in
                self?.drainPendingLiveActivityDecision()
            }
            .store(in: &cancellables)
    }

    private func drainPendingLiveActivityDecision() {
        guard let record = LiveActivitySharedStore.takePendingDecision() else { return }
        guard let current = prompt, current.id == record.id else {
            // Stale or already-answered prompt — just refresh the island.
            pushLiveActivity()
            return
        }
        let decision: PermissionDecision = record.decision == .approve ? .once : .deny
        respondPermission(decision)
    }

    func start(displayName: String? = nil, includeServiceUUIDInAdvertisement: Bool = true) {
        guard !started else { return }
        let finalName = resolvedDisplayName(displayName)
        activeDisplayName = finalName
        peripheral.setAdvertisementMode(includeServiceUUID: includeServiceUUIDInAdvertisement)
        peripheral.start(displayName: finalName)
        recordEvent("系统 请求启动广播：\(finalName)")
        refreshFromRuntime()
        started = true
        pushLiveActivity()
    }

    /// Called by onboarding to show the Bluetooth permission sheet after an
    /// explicit user tap (no auto-prompt at launch).
    func requestBluetoothAuthorization() {
        peripheral.requestAuthorization()
    }

    func stop() {
        guard started else { return }
        peripheral.stop()
        recordEvent("系统 BLE 外设已停止")
        started = false
        Task {
            await liveActivityManager.end()
        }
    }

    @discardableResult
    func respondPermission(_ decision: PermissionDecision) -> TimeInterval? {
        let answeredId = prompt?.id
        guard let line = runtime.respondPermission(decision) else { return nil }
        let direction = decision == .once ? "允许" : "拒绝"
        recordEvent("发送  权限\(direction)")
        _ = peripheral.sendLine(line)

        let elapsed = lastPromptAt.map { Date().timeIntervalSince($0) }
        switch decision {
        case .once:
            statsStore.onApproval(secondsToRespond: elapsed ?? 0)
            if let elapsed, elapsed < quickApprovalThreshold {
                lastQuickApprovalAt = Date()
            }
        case .deny:
            statsStore.onDenial()
        }
        lastPromptAt = nil
        lastPromptId = nil
        responseSent = true
        // Optimistically clear the prompt and refresh the Live Activity so the
        // Dynamic Island / lock-screen buttons collapse immediately instead of
        // waiting ~one heartbeat for the desktop to reflect the response. The
        // runtime is now idempotent against the same id, so a stale heartbeat
        // won't re-seat it. Also sweep any delivered local notification for
        // this prompt id — its quick actions would otherwise reappear.
        prompt = nil
        if let answeredId {
            BuddyNotificationCenter.shared.clearPromptNotifications(promptID: answeredId)
        }
        pushStatusSample()
        pushLiveActivity()
        return elapsed
    }

    /// Wipes the log surfaces visible in the home log sheet — heartbeat
    /// entries, BLE wire events, and diagnostic logs. Remote sources
    /// (heartbeat.entries) will refill on the next desktop tick; the intent is
    /// to give the user a clean slate, not to permanently suppress logs.
    func clearLogs() {
        recentEvents.removeAll()
        diagnosticLogs.removeAll()
        parsedEntries.removeAll()
        snapshot.entries.removeAll()
    }

    /// Records a user-visible device-menu interaction (MENU / SETTINGS /
    /// RESET apply) into the recent event log. Menu state itself is local —
    /// this is the only surface the user has to see what changed.
    func logDeviceMenuEvent(_ description: String) {
        recordEvent("设备 \(description)")
    }

    /// Incoming heartbeat `entries` arrive newest-first. Walk from oldest to
    /// newest and prepend any line that isn't already in our recent window —
    /// lines that stay in the bridge's top-N across several heartbeats get
    /// skipped instead of duplicated. Keeping the dedup to a window (not the
    /// whole history) means a legitimate repeat later on still records.
    private func mergeParsedEntries(from incoming: [String]) {
        guard !incoming.isEmpty else { return }
        for entry in incoming.reversed() {
            let window = parsedEntries.prefix(Self.parsedEntriesDedupWindow)
            if window.contains(entry) { continue }
            parsedEntries.insert(entry, at: 0)
        }
        if parsedEntries.count > Self.parsedEntriesMax {
            parsedEntries.removeLast(parsedEntries.count - Self.parsedEntriesMax)
        }
    }

    private func refreshFromRuntime() {
        let newSnapshot = runtime.currentSnapshot()
        mergeParsedEntries(from: newSnapshot.entries)
        snapshot = newSnapshot
        transfer = runtime.transferProgress()

        let newPrompt = runtime.pendingPrompt()
        if let newPrompt, newPrompt.id != lastPromptId {
            lastPromptId = newPrompt.id
            lastPromptAt = Date()
            responseSent = false
            BuddyNotificationCenter.shared.notifyPromptIfNeeded(
                promptID: newPrompt.id,
                tool: newPrompt.tool,
                enabled: notificationsEnabled
            )
        } else if newPrompt == nil {
            lastPromptAt = nil
            lastPromptId = nil
            responseSent = false
        }
        prompt = newPrompt

        let leveledUp = statsStore.onBridgeTokens(newSnapshot.tokens)
        if leveledUp {
            recentLevelUp = true
            recordEvent("系统 升级！等级 \(statsStore.stats.level)")
            BuddyNotificationCenter.shared.notifyLevelUpIfNeeded(
                level: UInt16(statsStore.stats.level),
                enabled: notificationsEnabled
            )
            Task { @MainActor [weak self] in
                try? await Task.sleep(for: .seconds(3))
                self?.recentLevelUp = false
            }
        }
        pushStatusSample()
        pushLiveActivity()
    }

    /// Single chokepoint for Live Activity updates. Honours the user's
    /// `buddy.liveActivityEnabled` toggle and the "only show when connected"
    /// rule — anything else tears the activity down instead of updating it.
    private func pushLiveActivity() {
        guard liveActivityEnabled, case .connected = connectionState else {
            Task { await liveActivityManager.end() }
            return
        }
        let slug = currentPersonaSlug(connection: connectionState, snapshot: snapshot)
        let preview = prompt.map(previewFor(prompt:))
        let state = connectionState
        let snap = snapshot
        let hasPrompt = prompt != nil
        let promptID = prompt?.id
        Task {
            await liveActivityManager.startOrUpdate(
                state: state,
                snapshot: snap,
                hasPrompt: hasPrompt,
                personaSlug: slug,
                messagePreview: preview,
                promptID: promptID
            )
        }
    }

    private func currentPersonaSlug(connection: NUSConnectionState, snapshot: BridgeSnapshot) -> String {
        let connected: Bool
        if case .connected = connection { connected = true } else { connected = false }
        let input = PersonaDeriveInput(
            connected: connected,
            sessionsRunning: snapshot.running,
            sessionsWaiting: snapshot.waiting,
            recentlyCompleted: recentLevelUp
        )
        return derivePersonaState(input).slug
    }

    private func previewFor(prompt: PromptRequest) -> String {
        let tool = prompt.tool.trimmingCharacters(in: .whitespacesAndNewlines)
        let hint = prompt.hint.trimmingCharacters(in: .whitespacesAndNewlines)
        if tool.isEmpty && hint.isEmpty { return String(localized: "live.alert.body") }
        if tool.isEmpty { return hint }
        if hint.isEmpty { return tool }
        return "\(tool): \(hint)"
    }

    private func enableBatteryMonitoring() {
        UIDevice.current.isBatteryMonitoringEnabled = true
        NotificationCenter.default.publisher(for: UIDevice.batteryLevelDidChangeNotification)
            .merge(with: NotificationCenter.default.publisher(for: UIDevice.batteryStateDidChangeNotification))
            .receive(on: DispatchQueue.main)
            .sink { [weak self] _ in self?.pushStatusSample() }
            .store(in: &cancellables)
        pushStatusSample()
    }

    private func pushStatusSample() {
        let device = UIDevice.current
        let rawLevel = device.batteryLevel
        let percent: Int = rawLevel < 0 ? 100 : Int((rawLevel * 100).rounded())
        let usb = (device.batteryState == .charging || device.batteryState == .full)
        let battery = BatterySample(percent: percent, millivolts: 0, milliamps: 0, usb: usb)
        let s = statsStore.stats
        let stats = StatsSample(
            approvals: Int(s.approvals),
            denials: Int(s.denials),
            velocityMedianSeconds: Int(s.medianVelocitySeconds),
            napSeconds: Int(s.napSeconds),
            level: Int(s.level)
        )
        runtime.updateStatusSample(StatusSample(battery: battery, stats: stats))
    }

    private func recordEvent(_ line: String) {
        recentEvents.insert(line, at: 0)
        if recentEvents.count > 120 {
            recentEvents.removeLast(recentEvents.count - 120)
        }
    }

    private func resolvedDisplayName(_ requested: String?) -> String {
        // OpenVibble Desktop 做大小写无关前缀匹配（"claude"），所以广播用
        // 全小写 "claude.openvibble" 让 OpenVibble 设备在扫描列表里可被一眼
        // 辨识。Claude Desktop 官方是大小写敏感的 "Claude" 前缀，要走它需要
        // 用户把 iPhone 系统名改成 Claude 开头后在后台运行（help sheet 说明）。
        _ = requested
        return "claude.openvibble"
    }

    private var notificationsEnabled: Bool {
        if UserDefaults.standard.object(forKey: "buddy.notificationsEnabled") == nil {
            return true
        }
        return UserDefaults.standard.bool(forKey: "buddy.notificationsEnabled")
    }

    private var liveActivityEnabled: Bool {
        if UserDefaults.standard.object(forKey: "buddy.liveActivityEnabled") == nil {
            return true
        }
        return UserDefaults.standard.bool(forKey: "buddy.liveActivityEnabled")
    }
}

extension Notification.Name {
    static let liveActivityDecisionReceived = Notification.Name("kingcos.me.openvibble.liveActivityDecisionReceived")
}

File: android/app/src/main/kotlin/com/openvibble/bridge/BridgeAppModel.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.openvibble.bridge

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.openvibble.nusperipheral.BuddyPeripheralService
import com.openvibble.nusperipheral.NusConnectionState
import com.openvibble.persona.PersonaSelectionStore
import com.openvibble.persona.PersonaSpeciesCatalog
import com.openvibble.persona.PersonaSpeciesId
import com.openvibble.protocol.PermissionDecision
import com.openvibble.runtime.BatterySample
import com.openvibble.runtime.BridgeRuntime
import com.openvibble.runtime.BridgeSnapshot
import com.openvibble.runtime.ProjectSummary
import com.openvibble.runtime.ProjectSummaryBuilder
import com.openvibble.runtime.PromptRequest
import com.openvibble.runtime.StatsSample
import com.openvibble.runtime.StatusSample
import com.openvibble.settings.AppSettings
import com.openvibble.settings.SharedPreferencesPersonaSelectionStore
import com.openvibble.stats.PersonaStatsStore
import com.openvibble.storage.CharacterTransferStore
import com.openvibble.storage.TransferProgress
import java.io.File
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.launch

/**
 * Android parity with iOS `BridgeAppModel` (OpenVibbleApp/BridgeAppModel.swift).
 *
 * Glues [BridgeRuntime] and [BuddyPeripheralService] into a single observable
 * surface for Compose screens. LiveActivity code from the iOS model has no
 * Android analogue and is intentionally omitted.
 *
 * Lifecycle: owned by Compose via `viewModel()`; tied to the process because
 * the BLE foreground service keeps it alive across configuration changes.
 */
class BridgeAppModel(
    private val context: Context,
    private val runtime: BridgeRuntime,
    private val peripheral: BuddyPeripheralService,
    val statsStore: PersonaStatsStore,
    private val settings: AppSettings,
) : ViewModel() {

    // Re-expose runtime/peripheral state flows directly — UI just binds to these.
    val snapshot: StateFlow<BridgeSnapshot> = runtime.snapshot
    val prompt: StateFlow<PromptRequest?> = runtime.prompt
    val transfer: StateFlow<TransferProgress> = runtime.transferProgress
    val connectionState: StateFlow<NusConnectionState> = peripheral.connectionState
    val bluetoothStateNote: StateFlow<String> = peripheral.bluetoothStateNote
    val bluetoothPowerState: StateFlow<com.openvibble.nusperipheral.BluetoothPowerState> = peripheral.powerState
    val advertisingNote: StateFlow<String> = peripheral.advertisingNote
    val diagnosticLogs: StateFlow<List<String>> = peripheral.diagnostics

    private val _activeDisplayName = MutableStateFlow(DEFAULT_DISPLAY_NAME)
    val activeDisplayName: StateFlow<String> = _activeDisplayName.asStateFlow()

    private val _recentEvents = MutableStateFlow<List<String>>(emptyList())
    val recentEvents: StateFlow<List<String>> = _recentEvents.asStateFlow()

    private val _parsedEntries = MutableStateFlow<List<String>>(emptyList())
    val parsedEntries: StateFlow<List<String>> = _parsedEntries.asStateFlow()

    private val _lastInstalledCharacter = MutableStateFlow<String?>(null)
    val lastInstalledCharacter: StateFlow<String?> = _lastInstalledCharacter.asStateFlow()

    private val _recentLevelUp = MutableStateFlow(false)
    val recentLevelUp: StateFlow<Boolean> = _recentLevelUp.asStateFlow()

    private val _lastQuickApprovalAt = MutableStateFlow<Long?>(null)
    val lastQuickApprovalAt: StateFlow<Long?> = _lastQuickApprovalAt.asStateFlow()

    private val _lastCompletedAt = MutableStateFlow<Long?>(null)
    val lastCompletedAt: StateFlow<Long?> = _lastCompletedAt.asStateFlow()

    private val _responseSent = MutableStateFlow(false)
    val responseSent: StateFlow<Boolean> = _responseSent.asStateFlow()

    /**
     * Per-project grouping of heartbeat entries, same API as iOS
     * `BridgeAppModel.projects`. Lazily computed via a cold-then-hot flow so
     * INFO > CLAUDE can collect it without forcing every Home-screen
     * recomposition to rebuild.
     */
    val projects: StateFlow<List<ProjectSummary>> =
        combine(_parsedEntries, runtime.prompt) { entries, p ->
            ProjectSummaryBuilder.build(entries = entries, hasPrompt = p != null)
        }.stateIn(
            scope = viewModelScope,
            started = SharingStarted.Eagerly,
            initialValue = emptyList(),
        )

    val charactersRoot: File get() = runtime.charactersRoot

    private var started: Boolean = false
    private var lastPromptId: String? = null
    private var lastPromptAtMs: Long? = null

    /** Pluggable hook for BuddyNotificationCenter so the port stays decoupled. */
    interface NotificationsBridge {
        fun notifyPromptIfNeeded(promptId: String, tool: String, enabled: Boolean)
        fun notifyLevelUpIfNeeded(level: Int, enabled: Boolean)
        fun clearPromptNotifications(promptId: String)
    }
    var notifications: NotificationsBridge? = null

    private val batteryReceiver = object : BroadcastReceiver() {
        override fun onReceive(ctx: Context?, intent: Intent?) { pushStatusSample() }
    }

    init {
        runtime.onCharacterInstalled = { name ->
            _lastInstalledCharacter.value = name
            recordEvent("系统 已安装宠物：$name")
        }
        runtime.onSpeciesChanged = { idx ->
            val line = when {
                idx == PersonaSpeciesCatalog.GIF_SENTINEL -> "系统 切换宠物 → GIF"
                else -> PersonaSpeciesCatalog.nameAt(idx)?.let { "系统 切换宠物 → $it (idx=$idx)" }
            }
            if (line != null) recordEvent(line)
        }
        runtime.onTaskCompleted = { _lastCompletedAt.value = System.currentTimeMillis() }

        peripheral.onLineReceived = { line ->
            recordEvent("接收  $line")
            val outbound = runtime.ingestLine(line)
            refreshFromRuntime()
            for (response in outbound) {
                recordEvent("发送  ${response.trim()}")
                peripheral.sendLine(response)
            }
        }

        // Runtime snapshot mutations happen outside refreshFromRuntime (e.g.
        // from status commands). Mirror snapshot.entries into parsedEntries
        // whenever it changes so INFO > CLAUDE always has fresh content.
        viewModelScope.launch {
            runtime.snapshot.collect { snap ->
                mergeParsedEntries(snap.entries)
            }
        }

        enableBatteryMonitoring()
    }

    fun start(includeServiceUuidInAdvertisement: Boolean = true) {
        if (started) return
        val name = DEFAULT_DISPLAY_NAME
        _activeDisplayName.value = name
        peripheral.setAdvertisementMode(includeServiceUuidInAdvertisement)
        peripheral.start(name)
        recordEvent("系统 请求启动广播：$name")
        refreshFromRuntime()
        started = true
    }

    fun stop() {
        if (!started) return
        peripheral.stop()
        recordEvent("系统 BLE 外设已停止")
        started = false
    }

    /** Returns elapsed ms since the prompt appeared, or null if no active prompt. */
    fun respondPermission(decision: PermissionDecision): Long? {
        val answeredId = runtime.prompt.value?.id
        val line = runtime.respondPermission(decision) ?: return null
        val direction = if (decision == PermissionDecision.ONCE) "允许" else "拒绝"
        recordEvent("发送  权限$direction")
        peripheral.sendLine(line)

        val elapsedMs = lastPromptAtMs?.let { System.currentTimeMillis() - it }
        when (decision) {
            PermissionDecision.ONCE -> {
                statsStore.onApproval(secondsToRespond = (elapsedMs ?: 0L).toDouble() / 1000.0)
                if (elapsedMs != null && elapsedMs < QUICK_APPROVAL_THRESHOLD_MS) {
                    _lastQuickApprovalAt.value = System.currentTimeMillis()
                }
            }
            PermissionDecision.DENY -> statsStore.onDenial()
        }
        lastPromptAtMs = null
        lastPromptId = null
        _responseSent.value = true
        answeredId?.let { notifications?.clearPromptNotifications(it) }
        pushStatusSample()
        return elapsedMs
    }

    fun clearLogs() {
        _recentEvents.value = emptyList()
        _parsedEntries.value = emptyList()
    }

    fun logDeviceMenuEvent(description: String) {
        recordEvent("设备 $description")
    }

    private fun mergeParsedEntries(incoming: List<String>) {
        if (incoming.isEmpty()) return
        val merged = _parsedEntries.value.toMutableList()
        for (entry in incoming.asReversed()) {
            val window = merged.take(PARSED_ENTRIES_DEDUP_WINDOW)
            if (window.contains(entry)) continue
            merged.add(0, entry)
        }
        if (merged.size > PARSED_ENTRIES_MAX) {
            merged.subList(PARSED_ENTRIES_MAX, merged.size).clear()
        }
        _parsedEntries.value = merged
    }

    private fun refreshFromRuntime() {
        val snap = runtime.snapshot.value
        mergeParsedEntries(snap.entries)

        val newPrompt = runtime.prompt.value
        if (newPrompt != null && newPrompt.id != lastPromptId) {
            lastPromptId = newPrompt.id
            lastPromptAtMs = System.currentTimeMillis()
            _responseSent.value = false
            notifications?.notifyPromptIfNeeded(
                promptId = newPrompt.id,
                tool = newPrompt.tool,
                enabled = settings.notificationsEnabled,
            )
        } else if (newPrompt == null) {
            lastPromptAtMs = null
            lastPromptId = null
            _responseSent.value = false
        }

        val leveled = statsStore.onBridgeTokens(bridgeTotal = snap.tokens.toLong())
        if (leveled) {
            _recentLevelUp.value = true
            val level = statsStore.stats.value.level
            recordEvent("系统 升级！等级 $level")
            notifications?.notifyLevelUpIfNeeded(level = level.toInt(), enabled = settings.notificationsEnabled)
            viewModelScope.launch {
                kotlinx.coroutines.delay(3_000L)
                _recentLevelUp.value = false
            }
        }
        pushStatusSample()
    }

    private fun enableBatteryMonitoring() {
        val filter = IntentFilter(Intent.ACTION_BATTERY_CHANGED)
        context.registerReceiver(batteryReceiver, filter)
        pushStatusSample()
    }

    private fun pushStatusSample() {
        val battery = readBattery() ?: BatterySample(percent = 100, usb = true)
        val s = statsStore.stats.value
        val sample = StatusSample(
            battery = battery,
            stats = StatsSample(
                approvals = s.approvals.toInt(),
                denials = s.denials.toInt(),
                velocityMedianSeconds = s.medianVelocitySeconds.toInt(),
                napSeconds = s.napSeconds.toInt(),
                level = s.level.toInt(),
            ),
        )
        runtime.updateStatusSample(sample)
    }

    private fun readBattery(): BatterySample? {
        val intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED)) ?: return null
        val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
        val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
        val percent = if (level < 0 || scale <= 0) 100 else ((level.toFloat() / scale.toFloat()) * 100f).toInt()
        val statusInt = intent.getIntExtra(BatteryManager.EXTRA_STATUS, BatteryManager.BATTERY_STATUS_UNKNOWN)
        val usb = statusInt == BatteryManager.BATTERY_STATUS_CHARGING || statusInt == BatteryManager.BATTERY_STATUS_FULL
        val millivolts = intent.getIntExtra(BatteryManager.EXTRA_VOLTAGE, 0)
        return BatterySample(percent = percent, millivolts = millivolts, milliamps = 0, usb = usb)
    }

    private fun recordEvent(line: String) {
        val current = _recentEvents.value.toMutableList()
        current.add(0, line)
        if (current.size > RECENT_EVENTS_MAX) {
            current.subList(RECENT_EVENTS_MAX, current.size).clear()
        }
        _recentEvents.value = current
    }

    override fun onCleared() {
        runCatching { context.unregisterReceiver(batteryReceiver) }
        stop()
        super.onCleared()
    }

    companion object {
        const val DEFAULT_DISPLAY_NAME: String = "claude.openvibble"
        const val RECENT_EVENTS_MAX: Int = 120
        const val PARSED_ENTRIES_MAX: Int = 200
        const val PARSED_ENTRIES_DEDUP_WINDOW: Int = 16
        const val QUICK_APPROVAL_THRESHOLD_MS: Long = 5_000L
    }

    class Factory(private val applicationContext: Context) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            require(modelClass.isAssignableFrom(BridgeAppModel::class.java)) {
                "Unsupported ViewModel: $modelClass"
            }
            val charactersRoot = File(applicationContext.filesDir, "characters").apply { mkdirs() }
            val transferStore = CharacterTransferStore(rootDirectory = charactersRoot)
            val personaSelection = SharedPreferencesPersonaSelectionStore(applicationContext)
            val runtime = BridgeRuntime(
                transferStore = transferStore,
                personaSelection = personaSelection,
            )
            val peripheral = BuddyPeripheralService(applicationContext)
            val statsStore = PersonaStatsStore()
            val settings = AppSettings(applicationContext)
            @Suppress("UNCHECKED_CAST")
            return BridgeAppModel(
                context = applicationContext,
                runtime = runtime,
                peripheral = peripheral,
                statsStore = statsStore,
                settings = settings,
            ) as T
        }
    }
}
